src/Tasks/SystemState.cs

```diff
@@ -540,16 +676,93 @@ out fileState.frameworkName
 
             dependencies = fileState.dependencies;
             scatterFiles = fileState.scatterFiles;
-            frameworkName = fileState.frameworkName;
+            frameworkName = fileState.FrameworkNameAttribute;
         }
 
         /// <summary>
-        /// Cached implementation of GetDirectories.
+        /// Reads in cached data from stateFiles to build an initial cache. Avoids logging warnings or errors.
         /// </summary>
-        /// <param name="path"></param>
-        /// <param name="pattern"></param>
-        /// <returns></returns>
-        private string[] GetDirectories(string path, string pattern)
+        internal static SystemState DeserializePrecomputedCaches(string[] stateFiles, TaskLoggingHelper log, Type requiredReturnType, GetLastWriteTime getLastWriteTime, AssemblyTableInfo[] installedAssemblyTableInfo)
+        {
+            SystemState retVal = new SystemState();
+            retVal.SetGetLastWriteTime(getLastWriteTime);
+            retVal.SetInstalledAssemblyInformation(installedAssemblyTableInfo);
+            retVal.isDirty = stateFiles.Length > 0;
+            HashSet<string> assembliesFound = new HashSet<string>();
+
+            foreach (string stateFile in stateFiles)
+            {
+                // Verify that it's a real stateFile; log message but do not error if not
+                var deserializeOptions = new JsonSerializerOptions() { Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping };
+                deserializeOptions.Converters.Add(new SystemState.Converter());
+                SystemState sysBase = JsonSerializer.Deserialize<SystemState>(File.ReadAllText(stateFile), deserializeOptions);
```

Comment:
Is json really faster than binary serialization? I thought the whole point of this feature is to maximize perf. Can you please do some measurements on real SystemState data to compare binary serialization vs json, so we're aware what the impact is?

Reply:
Most of the reason is that BinaryFormatter is unsafe, and we're supposed to move all of MSBuild off of it. I should measure perf at some point, but I haven't gotten around to it yet.

Reply:
If that's the case, we should be replacing it with another binary serialization library that's safe and as fast or faster.

Reply:
I just did a janky test comparing the time of:
```
[Fact]
        public void SpeedTest()
        {
            string data = File.ReadAllText(@"<path>");
            for (int i = 0; i < 1000; i++)
            {
                using (FileStream s = new FileStream(@"<path>", FileMode.Open))
                {
                    var formatter = new System.Runtime.Serialization.Formatters.Binary.BinaryFormatter();
                    SystemState t = formatter.Deserialize(s) as SystemState;
                }
            }
        }
```
minus the time of
```
[Fact]
        public void SpeedTest()
        {
            for (int i = 0; i < 1000; i++)
            {
                string data = File.ReadAllText(@"<path>");
            }
        }
```

to the time of

```
[Fact]
        public void TestSpeed()
        {
            string data = File.ReadAllText(@"<path>");
            var deserializeOptions = new JsonSerializerOptions() { Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping };
            deserializeOptions.Converters.Add(new SystemState.Converter());
            for (int i = 0; i < 1000; i++)
                JsonSerializer.Deserialize<SystemState>(data, deserializeOptions);
        }
```

For the json serializer, I used a statefile created by combining the state files cached when building all of MSBuild, whereas I used just the state file cache for Microsoft.Build.Engine.UnitTests for BinaryFormatter, so it should be smaller. Still, the two runs for the binary formatter took 2 seconds (unfortunately not more precise) and 435 ms, whereas json took 996 ms, suggesting that for this test at least, json was slightly faster.